Iter9

abezemskiy opened the pull request.

abezemskiy committed:
Merge branch 'main' of github.com:AntonBezemskiy/go-musthave-metrics

abezemskiy committed:
Refactor internal directory structure for agent and server code

abezemskiy committed:
Replace repositories interface implementation closer to source code

abezemskiy committed:
Replace repositories interface implementation closer to source code

abezemskiy committed:
Add logger for agent and server

abezemskiy committed:
Fix environment variable name of http address

abezemskiy committed:
Add an API for the server to receive metrics in json

abezemskiy committed:
Add API for the server to push metrics in json

abezemskiy committed:
Add API for the agent to push metrics in json

abezemskiy committed:
Fix repositories.Metrics variable declaration

abezemskiy committed:
Add gzip compession support

abezemskiy committed:
Add support saving metric into the file

abezemskiy committed:
Use encoder and decoder instead marshal and unmarshal

abezemskiy committed:
Fix test file path

abezemskiy committed:
Fix test, use GetCounters and GetGauges methods of storage.MemStorage structure

abezemskiy committed:
Remove buffering step, save metrics directly to file from server

tmvrus commented on cmd/server/main.go:
```diff
@@ -3,48 +3,89 @@ package main
 
-	if err != nil {
-		log.Printf("Error starting server: %v\n", err)
+	time.Sleep(100 * time.Millisecond)
```
а зачем это?

abezemskiy replied:
Осталось с предыдущих версий реализации, удалил в 4ecd677e696dba1350739d70617154f9f3ef3a38.

tmvrus commented on cmd/server/main.go:
```diff
@@ -3,48 +3,89 @@ package main
+		if err != nil {
+			logger.ServerLog.Error("flushing metrics error", zap.String("error", error.Error(err)))
+		}
+		time.Sleep(saver.GetStoreInterval() * time.Second)
```
`saver.GetStoreInterval() * time.Second` можно вычислить 1 раз перед входом в цикл

abezemskiy replied:
Исправил в 4ecd677e696dba1350739d70617154f9f3ef3a38.

tmvrus commented on internal/agent/compress/compress.go:
```diff
@@ -0,0 +1,61 @@
+	"fmt"
+)
+
+var ContentTypes []string
+
+func init() {
+	ContentTypes = []string{
+		"application/javascript",
```
а где это используется?

abezemskiy replied:
Добавил использование в 4ecd677e696dba1350739d70617154f9f3ef3a38.  
Перенес из internal/agent/compress/compress.go в internal/server/compress/compress.go.

tmvrus commented on internal/agent/handlers/handlers.go:
```diff
@@ -0,0 +1,181 @@
+	// сериализую полученную струтктуру с метриками в json-представление  в виде слайса байт
+	var bufEncode bytes.Buffer
+	enc := json.NewEncoder(&bufEncode)
+	if err := enc.Encode(metrics); err != nil {
+		logger.AgentLog.Error("Encode message error", zap.String("error", error.Error(err)))
+	}
```
потерян return

abezemskiy replied:
Добавил return в 4ecd677e696dba1350739d70617154f9f3ef3a38.

tmvrus commented on internal/agent/handlers/handlers.go:
```diff
@@ -0,0 +1,181 @@
+
+	for _, metricName := range storage.AllMetrics {
+		typeMetric, value, err := metrics.GetMetricString(metricName)
+		if err != nil {
+			logger.AgentLog.Error(fmt.Sprintf("Failed to get metric %s: %v\n", typeMetric, err), zap.String("action", "push metrics"))
+		}
+		er := PushJSON(address, action, typeMetric, metricName, value, client)
```
а что запушит этот код если была ошибка при  получении метрики?

abezemskiy replied:
Добавил continue, в случае возникновения ошибки при получении метрики в 4ecd677e696dba1350739d70617154f9f3ef3a38.

tmvrus commented on internal/server/saver/saver.go:
```diff
@@ -0,0 +1,274 @@
+	return nil
+}
+
+// // Сохраняю метрики в буфер для последующей записи в файл
```
закоменченого кода в готовом проекте быть не должно, это замусоривать и усложняет чтение

abezemskiy replied:
Удалил закомментированный код в 4ecd677e696dba1350739d70617154f9f3ef3a38  .
Надеюсь ничего, что все исправления в одном коммите сделал.

tmvrus reviewed (changes requested):
Хорошая работа, но есть несколько вопросов

abezemskiy committed:
Fix review comments:

remove unused timer, count sleep interval once, add using ContentTypes condition,
add return, omit pushing metric if get metric error occurred, remove commented code
